fix(snownet): implement STUN keepalive with relays

thomaseizinger opened the pull request:
Firezone Clients and Gateways create an allocation with a given set of Relays as soon as they start up. If no traffic is being secured and thus no connections are established between them, NAT bindings between Clients / Gateways and the Relays may expire. Typically, these bindings last for 120s. Allocations are only refreshed every 5 min (after 50% of their lifetime has passed). 

After a NAT binding is expired, the next UDP message passing through the NAT may allocate a new port, thus changing the 3-tuple of the sender. TURN identifies clients by their 3-tuple. Therefore, without a proactive keepalive, TURN clients lose access to their allocation and need to create one under the new port.

To fix this, we implement a scheduled STUN binding request every 25s once we have chosen a socket (IPv4 or IPv6) for a given relay.

Resolves: #7802.

thomaseizinger committed:
Send STUN keep-alives every 25s

jamilbk reviewed (commented):
Nice work on the quick turnaround! Taking a closer look, but before I do, just to clarify this part:

> It only affects idle gateways behind restrictive NATs.

I want to make sure I'm understanding:

Let's say the Client is behind symmetric NAT, and the Gateway is behind a NAT with a low UDP NAT binding lifetime (120s).

When the Client wishes to communicate to the Gateway, they end up settling on the following path:

```
client:50000 <-> 3478:relay:49152 <-> 60000:gateway
```

In the above, the client's TURN channel data messages arrive at the Relay's control port, and then are demuxed to its allocation port of 49152 onwards to the Gateway.

Now, the Client and Gateway are happily communicating, then at the 5-minute mark, the Gateway goes to refresh its allocation, upon which the Relay doesn't recognize it any longer and a new allocation must be made.

In the 5 minutes that data is happily flowing between Client and Gateway, has _any_ traffic been sent from the Gateway to the Relay's control port?

As far as the Gateway is concerned, it's communicating with the Client on the Relay's 49152 allocation port like it would any other nominated candidate, correct?

jamilbk reviewed (approved):
LGTM! Makes sense. If I understand correctly this timer is tied to the active socket per Relay, so they'll automatically be reset when roaming and removing/replacing Relays.

thomaseizinger committed:
Test that expired binding requests timeout the allocation

thomaseizinger committed:
Refactor invalidation logic

thomaseizinger committed:
Fix refresh after half-lifetime test

thomaseizinger committed:
Remove outdated test

At the time that this test was written, we called `refresh` every time
an allocation was meant to be used for a connection.

This is no longer the case so fixing this test has no value.

thomaseizinger committed:
Only send keep-alives if we have an allocation

jamilbk reviewed (approved):
LGTM. Couple questions to make sure I'm following:

1. Am I understanding this correctly that a single binding failure will invalidate the allocation? If so, is that what we want? 
2. Are binding indications worth crafting for this purpose instead to save a few bytes?

thomaseizinger committed:
Only disconnect on timeouts of the same IP version as the socket
